ASSERTION FAILED: owner->get() == &Thread::current() in JSC::BlockDirectory::assertIsMutatorOrMutatorIsStopped for inspector/memory/tracking.html

kmiller68 opened the pull request:
#### 89d69b59b62aa4ca20e50614efff5e1c34269291
<pre>
ASSERTION FAILED: owner-&gt;get() == &amp;Thread::current() in JSC::BlockDirectory::assertIsMutatorOrMutatorIsStopped for inspector/memory/tracking.html
<a href="https://bugs.webkit.org/show_bug.cgi?id=275351">https://bugs.webkit.org/show_bug.cgi?id=275351</a>
<a href="https://rdar.apple.com/problem/129580672">rdar://problem/129580672</a>

Reviewed by Yusuke Suzuki.

The assert in of `m_objectSpace.capacity() &gt;= m_objectSpace.size()` in `Heap::extraMemorySize()` is
invalid as calling `m_objectSpace.size()` requires being the mutator currently, which
`WebCore::ResourceUsageThread` is not.

* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::extraMemorySize):

Canonical link: <a href="https://commits.webkit.org/279920@main">https://commits.webkit.org/279920@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b913bf47d5cc6f76b1c5337fa31646360c2a60a4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54907 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34355 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7495 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58185 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5638 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41900 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5666 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44469 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3820 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57002 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32452 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47571 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25596 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29242 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4912 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3779 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/48275 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5134 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59775 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/54412 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30165 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5280 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51889 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31301 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47649 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51315 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed, /WebKitGTK/TestInspectorServer:/webkit/WebKitWebInspectorServer/http-test-page-list (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12075 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32316 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/66708 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31090 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12725 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/heap/Heap.cpp:
```diff
@@ -969,7 +969,8 @@ size_t Heap::extraMemorySize()
     size_t total = UNLIKELY(checkedTotal.hasOverflowed()) ? std::numeric_limits<size_t>::max() : checkedTotal.value();
 
-    ASSERT(m_objectSpace.capacity() >= m_objectSpace.size());
+    // It would be nice to have `ASSERT(m_objectSpace.capacity() >= m_objectSpace.size());` here but `m_objectSpace.size()`
```
Can we just check if we are the resource usage thread and skip the assert in that case? Or do you think it isn't worth it?

kmiller68 replied:
We could re-add the assert once we have a global GC thus will need the ability to share heap access anyway.

Constellation reviewed (approved):
r=me

kmiller68 committed:
ASSERTION FAILED: owner->get() == &Thread::current() in JSC::BlockDirectory::assertIsMutatorOrMutatorIsStopped for inspector/memory/tracking.html
https://bugs.webkit.org/show_bug.cgi?id=275351
rdar://problem/129580672

Reviewed by Yusuke Suzuki.

The assert in of `m_objectSpace.capacity() >= m_objectSpace.size()` in `Heap::extraMemorySize()` is
invalid as calling `m_objectSpace.size()` requires being the mutator currently, which
`WebCore::ResourceUsageThread` is not.

* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::extraMemorySize):

Canonical link: https://commits.webkit.org/279920@main
